engine/src/test/java/org/camunda/bpm/engine/test/bpmn/tasklistener/TaskListenerEventLifecycleTest.java

```diff
@@ -0,0 +1,586 @@
+    List<String> orderedEvents = RecorderTaskListener.getOrderedEvents();
+
+    assertThat(orderedEvents.size()).isEqualTo(3);
+    // ASSIGNMENT Event is fired, since the ModifyingTaskListener sets an assignee, and the
+    // ASSIGNMENT Event evaluation happens after the UPDATE Event evaluation
+    assertThat(orderedEvents).containsExactly(TaskListener.EVENTNAME_CREATE,
+                                              TaskListener.EVENTNAME_UPDATE,
+                                              TaskListener.EVENTNAME_ASSIGNMENT);
+  }
+
+  @Test
+  public void testNoUpdateEventAfterAssignmentTaskListenerUpdatesProperties() {
+    // given
+    BpmnModelInstance process = createModelWithTaskEventsRecorderOnAssignedUserTask(TRACKED_EVENTS,
+                                                                                    null,
+                                                                                    TaskListener.EVENTNAME_ASSIGNMENT,
+                                                                                    ModifyingTaskListener.class);
+    testRule.deploy(process);
+    engineRule.getRuntimeService().startProcessInstanceByKey("process");
+    Task task = engineRule.getTaskService().createTaskQuery().singleResult();
+
+    // when
+    taskService.setAssignee(task.getId(), "john");
+
+    // then
+    // only one update event is expected, from the initial assignment
+    List<String> orderedEvents = RecorderTaskListener.getOrderedEvents();
+
+    assertThat(orderedEvents.size()).isEqualTo(3);
+    assertThat(orderedEvents).containsExactly(TaskListener.EVENTNAME_CREATE,
+                                              TaskListener.EVENTNAME_UPDATE,
+                                              TaskListener.EVENTNAME_ASSIGNMENT);
+  }
+
+  @Test
+  public void testNoUpdateEventAfterCompleteTaskListenerUpdatesProperties() {
+    // given
+    BpmnModelInstance process = createModelWithTaskEventsRecorderOnAssignedUserTask(TRACKED_EVENTS,
+                                                                                    null,
+                                                                                    TaskListener.EVENTNAME_COMPLETE,
+                                                                                    ModifyingTaskListener.class);
+    testRule.deploy(process);
+    engineRule.getRuntimeService().startProcessInstanceByKey("process");
+    Task task = engineRule.getTaskService().createTaskQuery().singleResult();
+
+    // when
+    taskService.complete(task.getId());
+
+    // then
+    List<String> orderedEvents = RecorderTaskListener.getOrderedEvents();
+
+    assertThat(orderedEvents.size()).isEqualTo(2);
+    assertThat(orderedEvents).containsExactly(TaskListener.EVENTNAME_CREATE,
+                                              TaskListener.EVENTNAME_COMPLETE);
+  }
+
+  @Test
+  public void testNoUpdateEventAfterDeleteTaskListenerUpdatesProperties() {
```

Comment:
for new test classes we should try to stay consistent with the test method naming and stick to "shouldXYZ"
```suggestion
  public void shouldNotFireUpdateEventAfterDeleteTaskListenerUpdatesProperties() {
```